dialer.go

```diff
@@ -81,137 +83,161 @@ func NewDialer(opts *DialerOpts) Dialer {
+	addr                  string
 }
 
-func (d *dialer) Dial(dial DialFN) (net.Conn, error) {
-	return d.DialContext(context.Background(), dial)
+type pendingSession struct {
+	name         string
+	dialTimeout  time.Duration
+	sleepOnError time.Duration
 }
 
-func (d *dialer) DialContext(ctx context.Context, dial DialFN) (net.Conn, error) {
-	s, err := d.getOrCreateSession(ctx, dial)
-	if err != nil {
-		return nil, err
+func newPendingSession() *pendingSession {
+	return &pendingSession{
+		name:         "background",
+		dialTimeout:  defaultDialTimeout,
+		sleepOnError: 2 * time.Second,
 	}
-	c := s.CreateStream()
-	d.returnSession(s)
-	return c, nil
-}
-
-func (d *dialer) getNumLivePending() int {
-	d.muNumLivePending.Lock()
-	numLivePending := d.numLive + d.numPending
-	d.muNumLivePending.Unlock()
-	return numLivePending
 }
 
-func (d *dialer) getOrCreateSession(ctx context.Context, dial DialFN) (sessionIntf, error) {
-	newSession := func(cap int) {
-		d.muNumLivePending.Lock()
-		if d.numLive+d.numPending >= cap {
-			d.muNumLivePending.Unlock()
-			return
-		}
-		d.numPending++
-		d.muNumLivePending.Unlock()
-		go func() {
-			s, err := d.startSession(dial)
-			d.muNumLivePending.Lock()
-			d.numPending--
+// maintainTCPConnections maintains background TCP connection(s) and associated lampshade session(s)
+func (d *dialer) maintainTCPConnections() (net.Conn, error) {
+	for {
+		select {
+		case rs := <-d.pendingSessions:
+			start := time.Now()
+			s, err := d.startSession(rs)
 			if err != nil {
-				d.muNumLivePending.Unlock()
-				return
+				log.Debugf("Error starting session '%v': %v", rs.name, err.Error())
+				time.Sleep(rs.sleepOnError)
+				d.pendingSessions <- rs
+			} else {
+				log.Debugf("Created session in %v", time.Since(start))
+				d.liveSessions <- s
 			}
-			d.numLive++
-			d.muNumLivePending.Unlock()
-			d.liveSessions <- s
-		}()
+		}
+	}
+}
+
+func (d *dialer) Dial() (net.Conn, error) {
+	return d.DialContext(context.Background())
+}
+
+func (d *dialer) DialContext(ctx context.Context) (net.Conn, error) {
+	s, err := d.getSession(ctx)
+	if err != nil {
+		return nil, err
+	}
+	c := s.createStream(ctx)
+	select {
+	case d.liveSessions <- s:
+	default:
+		log.Debug("Maximum live sessions reached")
 	}
+
+	return c, nil
+}
+
+func (d *dialer) getSession(ctx context.Context) (sessionIntf, error) {
+	start := time.Now()
 	for {
 		select {
 		case s := <-d.liveSessions:
-			if s.AllowNewStream(d.maxStreamsPerConn, d.idleInterval) {
-				return s, nil
+			if s.isClosed() {
+				// Closed sessions will trigger the creation of new ones, so keep waiting for a new session.
+				continue
+			}
+
+			if !s.allowNewStream(d.maxStreamsPerConn) {
+				// The default number of streams per session is 65535, so this is unlikely to be reached.
+				select {
+				case d.pendingSessions <- newPendingSession():
+					log.Debug("Added required session")
+				default:
+					log.Debug("Maximum sessions reached")
+				}
```

Comment:
Since `s. allowNewStream` should rarely be `false`, it may be worth printing one log message explicitly, or, change the above debug prints to emphasize that it's the result of max streams per conn being hit.

Reply:
Shouldn't it `continue` to get the next live session? The current session can't create new streams anyway.

Reply:
Yes great catch on the continue!